multi: add client epoch queue.

dnldd opened the pull request:
This adds an epoch queue to the client with match proof validation. This comprises of generating the seed from the first 64-bits of sorted and concatenated epoch order preimages and generating the commitment checksum from the sorted and concatenated epoch order commitments.

This will be rebased and updated to use the types defined in #145 when that's ready.

buck54321 commented on client/order/epochqueue.go:
```diff
@@ -0,0 +1,126 @@
+
+// EpochQueue represents a client epoch queue.
+type EpochQueue struct {
+	orders map[string]*msgjson.EpochOrderNote
```
Use `order.OrderID` for keys.

buck54321 commented on client/order/epochqueue.go:
```diff
@@ -0,0 +1,126 @@
+	eq.orders[id] = note
+
+	// Update the queue sort order per order ids.
+	i := sort.Search(len(eq.index), func(i int) bool { return strings.Compare(eq.index[i], id) > 0 })
```
Looks like your sorting on order ID, but the spec says the checksum is generated from sorted commitment hashes.

chappjc commented on client/order/epochqueue.go:
```diff
@@ -0,0 +1,190 @@
+
+// GenerateMatchProof calculates the sorting seed used in order matching as well as the
+// commitment checksum from the provided epoch queue preimages and misses.
+func (eq *EpochQueue) GenerateMatchProof(preimages []order.Preimage, misses []string) (int64, msgjson.Bytes, error) {
```
Why a string instead of order.OrderID for misses?

dnldd replied:
The spec indicated a string slice. Will update to order.OrderID slice.

chappjc replied:
I suppose it depends on who is calling `GenerateMatchProof`, but I don't think it needs to know about the msgjson type where the spec comes in to play.  The json serializations for byte slices/arrays are hexadecimal-encoded strings, but perhaps the spec could use updates in spots to make this explicit everywhere.

buck54321 replied:
@dnldd the spec only describes protocols for communication. There is no byte arrays in JSON, so they are expressed as hex strings. But they are order IDs, and we should be reflecting that in our implementation.

chappjc commented on client/order/epochqueue.go:
```diff
@@ -0,0 +1,190 @@
+		delete(eq.orders, oid)
+
+		// Remove the seed index associated with the miss, preserving
+		// the sort order in the process.
```
The "index" terminology throws me off.  This comment is closer to "remove the order ID from the seed index..." because I believe you are using "index" in the sense of a catalog.  There are no indexes in the sense we usually use the term to mean an integer location in a slice or array, correct?

chappjc commented on client/order/epochqueue.go:
```diff
@@ -0,0 +1,190 @@
+
+	// Map the preimages received with their associated epoch order ids.
+	matches := make(map[order.OrderID]msgjson.Bytes)
+	for i := 0; i < len(preimages); i++ {
```
`for i := range preimages {`

chappjc commented on client/order/epochqueue.go:
```diff
@@ -0,0 +1,190 @@
+			commitment := blake256.Sum256(preimages[i][:])
+			if bytes.Equal(note.Commitment, commitment[:]) {
+				matches[oid] = preimages[i][:]
+				break
```
Perhaps `continue outerLoop` and then a warning after this inner loop because that would indicate an order for preimage X was not found.

chappjc commented on client/order/epochqueue.go:
```diff
@@ -0,0 +1,190 @@
+	}
+
+	// Map the preimages received with their associated epoch order ids.
+	matches := make(map[order.OrderID]msgjson.Bytes)
```
Might as well preallocate to len(preimages)

chappjc commented on client/order/epochqueue.go:
```diff
@@ -0,0 +1,190 @@
+	}
+	eq.seedIndexMtx.Unlock()
+
+	seed := int64(binary.LittleEndian.Uint64(sbuff[:64]))
```
The shuffle seed is also a hash of concatenated values.

chappjc commented on client/order/epochqueue.go:
```diff
@@ -0,0 +1,190 @@
+		eq.seedIndexMtx.Lock()
+		si := sort.Search(len(eq.seedIndex), func(i int) bool { return bytes.Compare(eq.seedIndex[i][:], oid[:]) >= 0 })
+		if si == len(eq.seedIndex) {
+			return 0, nil, fmt.Errorf("expected to find id %s in seed sort index", oid)
```
If these errors happen, the entire EpochQueue is dead, correct?  The `eq.orders` slice has already had this oid removed, so you will get panics when attempting to get Commitment by looping through `eq.commitmentIndex` for `csum` computation in future calls to this function.  I suppose that's OK but it should be clearly documented that if `GenerateMatchProof` returns a non-nil error, the `EpochQueue` is hosed and `Reset` should be called, etc.

dnldd replied:
Yes, for preimage mismatch errors and non existent orders for preimages the epoch queue should be reset.

chappjc commented on client/order/epochqueue.go:
```diff
@@ -0,0 +1,190 @@
+		eq.commitmentIndexMtx.Lock()
+		ci := sort.Search(len(eq.commitmentIndex), func(i int) bool { return bytes.Compare(eq.commitmentIndex[i][:], oid[:]) >= 0 })
+		if ci == len(eq.commitmentIndex) {
+			return 0, nil, fmt.Errorf("expected to find id %s in commitment sort index", oid)
```
Deadlock. `eq.commitmentIndexMtx` is still locked.

chappjc commented on client/order/epochqueue.go:
```diff
@@ -0,0 +1,190 @@
+			copy(eq.commitmentIndex[ci:], eq.commitmentIndex[ci+1:])
+		}
+		eq.commitmentIndex[len(eq.commitmentIndex)-1] = order.OrderID{}
+		eq.commitmentIndex = eq.commitmentIndex[:len(eq.commitmentIndex)-1]
```
I wonder if it's worth the complexity to keep a sorted list.  It only needs to be sorted in `GenerateMatchProof`, not any other time between calls to `Queue` as far as I can tell, so the order ids and commitments could be kept in maps, which would facilitate these Searches, and for the seed and csum computation, we could form slices and sort them since they are only needed in this function.

dnldd replied:
I think it's worth the trouble keeping track of the sort order for the epoch queue, `GenerateMatchProof` without needing to perform an order id sort and a commitment sort before generating the match proofs would be a lot faster. I think the current implementation would perform substantially better with larger epoch queues.

chappjc replied:
I think this deserves a bit more discussion.  There are several things to consider here:

1. Design goals: Unlike the order book, where keeping the list sorted on each new order received actually is a requirement, epoch validation does not need sorted lists until the epoch close. We do not get this for free; we pay in complexity and performance.

2. Complexity: Performance differences and performance impact (two distinct factors) have to be quite significant to justify additional complexity, which brings bugs and maintenance challenges. A small difference in performance does not justify a notable increase in complexity, and a large performance difference does not justify it either if both solutions have *acceptable* performance (e.g. 10ms vs 20ms in this case I might argue, although we should characterize this).

3. Scalability: The complexity vs. performance trade-off above is much harder to justify for the client, whereas server-side performance is a great concern.  The client has far less to worry about in terms of performance than the server, which has many more roles (e.g. swap coordination and numerous markets), may of which are time critical.

4. Performance: The statement that maintaining sorted lists on each new order insertion "would perform substantially better with larger epoch queues" is not well qualified, and, I believe, is incorrect.

To keep a list sorted, receiving N orders over the course of an epoch involves:
 - N slice location searches, which involves order ID comparisons in a binary search.
 - N copies of up to the entire slice to insert (with the slice growing from length from 0 to N over the epoch).
 
 To sort at epoch close:
 - N appends (no comparisons, and no copies except for occasional reallocations, which the insertions also require in addition to many copies on EACH insertion).
 - 1 sort on epoch close (swaps, no reallocations).  You will find that this is extremely fast if we are sorting a slice of pointers or short arrays.

I intend to benchmark a simple append-and-later-sort approach to compare with the currently proposed solution, and I encourage you to do the same.  But, there would have to be a considerable (and surprising to me) difference in performance to justify the increase in complexity and all that goes with it.

dnldd replied:
Alright, your perspective on the issue is a lot clearer to me now. I'll implement the alternative and benchmark as suggested.

dnldd replied:
Looks like there isn't much being gained from maintaining sort orders on enqueue. `BenchmarkMatchProofAlt`, which generates the sort orders when `GenerateMatchProof` is called is performing at about the same level as `BenchmarkMatchProof` which maintains sort orders. In the current `BenchmarkMatchProofAlt` I'm copying the order id and epoch orders from the order map into slices and sorting that using the order ids and order commitments. 

```
BenchmarkMatchProofAlt500-8           27          41963687 ns/op          414722 B/op       2556 allocs/op
BenchmarkMatchProofAlt1000-8           7         162401527 ns/op          827794 B/op       5064 allocs/op
BenchmarkMatchProofAlt5000-8           1        4149027722 ns/op         4740384 B/op      25170 allocs/op
BenchmarkMatchProof500-8              30          43055826 ns/op          393997 B/op       2549 allocs/op
BenchmarkMatchProof1000-8              7         160078052 ns/op          787010 B/op       5058 allocs/op
BenchmarkMatchProof5000-8              1        4092102140 ns/op         4538520 B/op      25172 allocs/op
```
`BenchmarkMatchProofAlt` does indeed has a simpler implementation, should be preferred as a result. Should I go ahead and make modifications or should I push the pr with the benchmarks?

chappjc replied:
Yes, I think a PR with the benchmarks would be interesting.

chappjc replied:
I was doing some testing and discovered that searching for an `OrderID` in `eq.commitmentIndex` does not work with `sort.Search` because the slice is not sorted by `OrderID`.

```go
ci := sort.Search(len(eq.commitmentIndex), func(i int) bool { return bytes.Compare(eq.commitmentIndex[i][:], oid[:]) >= 0 })
```
 can fail to find `oid`.

chappjc replied:
With a single map and sorting in GenerateMatchProof, here's what I get for a inserting 500/1000/5000 orders and getting the match proof with N/20 misses:

```
BenchmarkEpochQueue500-32     	33	  34366184 ns/op	   298102 B/op	    60 allocs/op
BenchmarkEpochQueue1000-32    	 8      135353264 ns/op	   593462 B/op	    84 allocs/op
BenchmarkEpochQueue5000-32    	 1	3331479770 ns/op	  2454240 B/op	   203 allocs/op
```

I couldn't do the same because of the `commitmentIndex` issue I mentioned in my previous comment.

dnldd replied:
Hit the issue you mentioned with the commitmentIndex, not sure why I thought I could sort the orders with their commitments and index the order ids instead. I'll need to lookup the the orders associated with indexed commitments to maintain the sorting. That should make things slower for the implementation. It's clear to me the complexity involved with maintaining the sorts doesn't provide the performance benefits I thought it would. I'm stripping it out.

chappjc commented on dex/msgjson/types.go:
```diff
@@ -578,9 +581,10 @@ type TradeNote struct {
+	Seq        uint64 `json:"seq,omitempty"`      // May be empty when part of an OrderBook.
+	MarketID   string `json:"marketid,omitempty"` // May be empty when part of an OrderBook.
+	OrderID    Bytes  `json:"oid"`
+	Commitment Bytes  `json:"com"`
```
I'll update the spec in https://github.com/decred/dcrdex/pull/154

chappjc replied:
Ah, I actually put the commitment in [`EpochOrderNote`](https://github.com/decred/dcrdex/pull/145/files#diff-9a36b4e081b2c1cc9032524815e5d57cR598) since the commitment only seems relevant there (not book/unbook).

chappjc commented on dex/msgjson/types.go:
```diff
@@ -443,6 +443,7 @@ type Coin struct {
 type Prefix struct {
 	signable
 	AccountID     Bytes  `json:"accountid"`
+	Commitment    Bytes  `json:"com"`
```
Despite the spec, I've put Commitment after ServerTime, and updated the spec in https://github.com/decred/dcrdex/pull/154

dnldd commented on client/order/epochqueue.go:
```diff
@@ -0,0 +1,190 @@
+
+// EpochQueue represents a client epoch queue.
+type EpochQueue struct {
+	orders             map[order.OrderID]*msgjson.EpochOrderNote
```
You're right, after taking another look it looks like only the commitment and the order id are are needed by the queue.

chappjc commented on client/order/epochqueue.go:
```diff
@@ -0,0 +1,190 @@
+	return &EpochQueue{
+		orders:          make(map[order.OrderID]*msgjson.EpochOrderNote),
+		seedIndex:       make([]order.OrderID, 0),
+		commitmentIndex: make([]order.OrderID, 0),
```
Unless there's a good reason, which should be documented, these slices don't need init since appends are fine without it.

chappjc commented on client/order/epochqueue.go:
```diff
@@ -0,0 +1,190 @@
+	seedIndex          []order.OrderID
+	seedIndexMtx       sync.Mutex
+	commitmentIndex    []order.OrderID
+	commitmentIndexMtx sync.Mutex
```
Are there any cases where either of `seedIndexMtx` or `commitmentIndexMtx` would be locked but **not** `ordersMtx          sync`?  I suspect you could eliminate the two index mutexes.

This would simplify Reset as well.  e.g.

```go
func (eq *EpochQueue) Reset() {
	eq.ordersMtx.Lock()
	eq.orders = make(map[order.OrderID]*msgjson.EpochOrderNote)
	eq.seedIndex = nil
	eq.commitmentIndex = nil
	eq.ordersMtx.Unlock()
}
```

chappjc commented on client/order/epochqueue.go:
```diff
@@ -0,0 +1,190 @@
+	eq.seedIndexMtx.Lock()
+	si := sort.Search(len(eq.seedIndex), func(i int) bool {
+		order := eq.orders[eq.seedIndex[i]]
+		return bytes.Compare(order.OrderID[:], oid[:]) > 0
```
As we discussed in Matrix chat, it looks like `eq.orders` is not needed here.  i.e. `eq.orders[eq.seedIndex[i]].OrderID == eq.seedIndex[i]`, which simplifies the insertion location search.

chappjc commented on client/order/epochqueue_test.go:
```diff
@@ -207,4 +287,86 @@ func TestEpochQueue(t *testing.T) {
+			commitments[i] = commitment
+		}
+
+		_, _, err := eq.GenerateMatchProofAlt(preimages, nil)
```
But this benchmark still uses the Enqueue with the unneeded operations.  A correct comparison needs two different EpochQueue types.
Also, the test data generation should not be part of the bench mark.  e.g. https://github.com/chappjc/dcrdex/commit/1bb44d76c801981bc4b7bc03bf8e78b55fcffca9#diff-e17d0f6ebad116a19390b508c97ef4f5R65-R96

dnldd replied:
I added `EnqueueAlt` for that but forgot use it in the benchmark.

dnldd commented:
@chappjc you were right the complexity in maintaining sorted order ids and commitments on queue didn't yield performance gains from my tests, I've simplified the queue as you suggested, updated the tests and benchmarks. Some feedback when you can, thanks.

chappjc commented on client/order/epochqueue.go:
```diff
@@ -0,0 +1,165 @@
+	for _, commit := range commitments {
+		cbuff = append(cbuff, commit[:]...)
+	}
+	csum := blake256.Sum256(cbuff)
```
We may consider creating a hasher to avoid actually concatenating into a large array, but using a new hasher ends up with more allocations, so this is OK for now.

i.e.
```go
hasher := blake256.New()
for i := range oids {
	pimg := orderPreimages[oids[i]]
	hasher.Write(pimg[:])
}
seed := hasher.Sum(nil)
```

Just something to consider if we don't want a very large `sbuff` and `cbuff` when order queues are large.

buck54321 commented on client/order/epochqueue.go:
```diff
@@ -0,0 +1,165 @@
+
+// EpochQueue represents a client epoch queue.
+type EpochQueue struct {
+	orders map[order.OrderID]order.Commitment
```
`Commitment` is all that's needed right now, but we will need order type, side, and value when implemented. See https://github.com/decred/dcrdex/blob/f981df6e9905bb7aa68a008cb87c9a9a1a98be6d/client/core/types.go#L131-L141.

chappjc commented on client/order/epochqueue.go:
```diff
@@ -0,0 +1,165 @@
+	}
+
+	sum := blake256.Sum256(sbuff)
+	seed := int64(binary.LittleEndian.Uint64(sum[:8]))
```
Oh, I missed this.  The seed is the hash, not the first 8 bytes of it as an int64.  So return `seed`, `[]byte` is used to hold the "32-byte number" described by the spec.

chappjc replied:
Ah, I need to update the spec as we used a 64-bit number before but a more appropriate seed is an array.  That really should have been updated when I did the mt19937 change, sorry.

chappjc replied:
Wow, I thought I updated the spec for the shuffle change, but I did not.  On my todo list.
